src/XrdOssStats/XrdStatsFileSystem.cc

```diff
@@ -0,0 +1,337 @@
+
+#include "XrdOuc/XrdOucGatherConf.hh"
+#include "XrdStatsConfig.hh"
+#include "XrdStatsDirectory.hh"
+#include "XrdStatsFile.hh"
+#include "XrdStatsFileSystem.hh"
+#include "XrdSys/XrdSysPthread.hh"
+#include "XrdXrootd/XrdXrootdGStream.hh"
+
+#include <inttypes.h>
+#include <stdexcept>
+#include <thread>
+
+StatsFileSystem::StatsFileSystem(XrdOss *oss, XrdSysLogger *lp, const char *configfn, XrdOucEnv *envP) :
+    m_oss(oss),
+    m_env(envP),
+    m_log(lp, "fsstat_"),
+    m_slow_duration(std::chrono::seconds(1))
+{
+    if (!oss) {
+        throw std::runtime_error("The storage statistics plugin must be chained with another filesystem.");
+    }
+    m_log.Say("------ Initializing the storage statistics plugin.");
+    if (!Config(configfn)) {
+        throw std::runtime_error("Failed to configure the storage statistics plugin.");
+    }
+    pthread_t tid;
+    int rc;
+    if ((rc = XrdSysThread::Run(&tid, StatsFileSystem::AggregateBootstrap, static_cast<void *>(this), 0, "FS Stats Compute Thread"))) {
+      m_log.Emsg("StatsFileSystem", rc, "create stats compute thread");
+      throw std::runtime_error("Failed to create the statistics computing thread.");
+    }
+    if (envP) {
+       m_gstream = reinterpret_cast<XrdXrootdGStream*>(envP->GetPtr("oss.gStream*"));
```

Comment:
I would really move the extraction of the gStream into the AddStorageSystem() method and return failure if you cannot get a gStream. Nothing worse than collecting statistics and not being able to report them. If you don't want to fail, then you could simply issue a message and return the current_ossp pointer and effectively you have disabled yourself with no harm done. You can't do any of that here.

Reply:
Done in the latest push.